python-checks/src/main/java/org/sonar/python/checks/UndeclaredNameUsageCheck.java

```diff
@@ -47,9 +55,38 @@ public void initialize(Context context) {
     context.registerSyntaxNodeConsumer(Tree.Kind.NAME, ctx -> {
       Name name = (Name) ctx.syntaxNode();
       if (!callGlobalsOrLocals && !hasWildcardImport && name.isVariable() && name.symbol() == null) {
-        ctx.addIssue(name, "Change its name or define it before using it");
+        ctx.addIssue(name, name.name() + " is not defined. Change its name or define it before using it");
       }
     });
+
+    context.registerSyntaxNodeConsumer(Tree.Kind.FUNCDEF, ctx -> {
+      FunctionDef functionDef = (FunctionDef) ctx.syntaxNode();
+      ControlFlowGraph cfg = ControlFlowGraph.build(functionDef, ctx.pythonFile());
+      if (cfg == null) {
+        return;
+      }
+      DefinedVariablesAnalysis analysis = DefinedVariablesAnalysis.analyze(cfg, functionDef.localVariables());
+      cfg.blocks().forEach(block -> checkCfgBlock(block, ctx, analysis.getDefinedVariables(block)));
+    });
+  }
+
+  private static void checkCfgBlock(CfgBlock cfgBlock, SubscriptionContext ctx, DefinedVariablesAnalysis.DefinedVariables definedVariables) {
+    Map<Symbol, DefinedVariablesAnalysis.VariableDefinition> currentState = new HashMap<>(definedVariables.getIn());
+    for (Tree element : cfgBlock.elements()) {
+      definedVariables.getVariableUsages(element).forEach((symbol, symbolUsage) -> {
+        if (symbolUsage.isWrite()) {
+          currentState.put(symbol, DefinedVariablesAnalysis.VariableDefinition.DEFINED);
+        }
+        DefinedVariablesAnalysis.VariableDefinition varDef = currentState.getOrDefault(symbol, DefinedVariablesAnalysis.VariableDefinition.BOTTOM);
```

Comment:
What's the point of `VariableDefinition.BOTTOM`? Can it happen that `currentState.get(symbol) == VariableDefinition.BOTTOM`?